README
Kernel Socks Bouncer for 2.6.x kernels


0] HOW ksb26 WORKS
1] HOW TO INSTALL ksb26
2] CONCLUSION

0] HOW ksb26 WORKS and WHY A KERNEL MODULE? 


* ksb26 redirects connections through a chain of SOCKS5 servers to improve your 
  anonimity on the Net.

* I have choosen this implementation to:
	* add support to ALL userspace programs that use kernel's sys_connect system call.
	* enjoy myself writing in kernel space.


1] HOW TO INSTALL ksb26

1.0] Installation script [ Recommended ]

Default path of the Linux kernel source directory is:
kernel_dir=/usr/src/linux
Default path for the installation of ksb26manager [userspace tool] is:
bin_dir=/usr/bin
You can modify this two path simply editing install.sh
To install ksb26 simply run ./install.sh 
This script: 
* builds ksb26 Linux kernel module and copies it to
/lib/modules/2.6.x/extra
* copies ksb26manager, the managment tool, to /usr/bin [default]

1.1] Manual installation

* Unpack ksb26-x-x-x.tar.gz archive:
~# mv ksb26-x-x-x.tar.gz /usr/src
~# cd /usr/src
~# tar zxvf ksb26.tar.gz

* Build ksb26.c running makemod.sh shell script:
~# cd /usr/src/ksb26/kernel/
~# ./makemod.sh

* Copy ksb26.ko to /lib/modules/2.6.x/extra [where 2.6.x is the version
of the Linux Kernel you're using].

* Copy ksb26manager [ksb26 Manager] to /usr/bin
[For more informations about ksb26manager read 1.4]

1.2] Loading ksb26 module and /dev/ksb26 device creation
~# modprobe ksb26 // loads the module with default values
~# modprobe ksb26 nsocks=N // loads the module; ksb26 will use a chain
				of N SOCKS [Ex. nsocks=3].
[~# insmod /lib/modules/2.6.x/extra/ksb26.ko ]

* Create a character device in /dev directory using major number of ksb26 kernel module.
To detect major number read dmesg kernel logs running for example:
~# dmesg
or
~# cat /var/log/messages
Between messages of ksb26 module there's this string:
[ksb26] Major device number = num
~# mknod /dev/ksb26 c num 0 // creates the char device

1.3] /dev/ksb26 device
/dev/ksb26 is the character device used by ksb26 to keep SOCKS and target hosts lists updated.
Strings formats accepted:
* ip addresses and port of SOCKS5
* ip addresses of target hosts [when ksb26 finds a new connection from localhost
to that host, this connection will pass through a SOCKS chain before reach the target host]
* clear SOCKS list string [ csl ] : ksb26 clears the SOCKS list

Socks format :
S=socks_ip:socks_port

Manual submission of a new SOCKS: ~# echo "S=127.0.0.1:1080" > /dev/ksb26

If you use ksb26manager you can set a default socks list in /etc/ksb26/socks that will be loaded
automatically by the userspace tool .

If you don't use ksb26manager you can create a new file as this below:
example:
S=127.0.0.1:1080
and then:
~# cat socks.txt > /dev/ksb26

To add a new target host to the list [if you use the userspace manager see /etc/ksb26/thosts]:
H=host_ip:host_port
If host_ip is *, connection directed to all IPs will be redirected.
If host_port is 0, all connection directed to that host will be redirected through a Socks chain.
example:
H=*:22
H=192.168.0.1:22
H=192.168.0.2:0
H=*:0 [deprecated]

To clear SOCKS list:
C

1.4] ksb26manager
is the userspace manager for ksb26 lkm. It needs wget, egrep and sed to work.
[You don't need ksb26manager. You can add SOCKS and target hosts manually via /dev/ksb26 character device]
ksb26manager keeps updated the SOCKS list using http://ksb26.sourceforge.net/socks/socks5.php

ksb26manager [-h] [-d ksb26_device] [-m ksb26_lkm_name] [-n nsocks] [-t thosts_file] [-w wait_time]
[-d ksb26_device]: character device used by ksb26
[-m ksb26_lkm_name]: lkm's name to use in modprobe command
[-n nsocks]: number of SOCKS used in chains
[-t thosts_file]: file containing target hosts [default /etc/ksb26/thosts]
[-s socks_file]: file containing default socks list [default /etc/ksb26/thosts]
[-w wait_time]: number of seconds to wait before updating the SOCKS list
[-h]: help

2] CONCLUSION
For any question, problem or request write me to:
paolo.ardoino@gmail.com


ChangeLog
2008-01-29  Paolo Ardoino <paolo.ardoino@gmail.com>
	* ksb26-0.0.5 for Linux v2.6.x
	* Various fix and changes. (tested on linux-2.6.23.12) 

2005-11-19  Paolo Ardoino <ardoino.gnu@disi.unige.it>
	* ksb26-0.0.4 for Linux v2.6.x

2005-11-19  Paolo Ardoino <ardoino.gnu@disi.unige.it>
	* socks.example: new file socks.example containing default socks.
	* install.sh: fixed thosts.example control, added installation
	of socks.example.
	* ksb26.8.gz: updated man page.


2005-09-29  Paolo Ardoino <ardoino.gnu@disi.unige.it>
	* ksb26-0.0.3 for Linux v2.6.x

2005-08-09  Paolo Ardoino  <ardoino.gnu@disi.unige.it>
	* kernel/ksb26*: added feature to redirect connections directed to all
	IPs [#H*:22;| redirects all connections directed to port 22].
	Read man page!

2005-06-09  Paolo Ardoino  <ardoino.gnu@disi.unige.it>
	* ksb26-0.0.2 for Linux v2.6.x
	* kernel/ksb26*: added feature to redirect connections directed to
	specific target host's ports [read manual].
	* user/ksb26manager.c: automatically detects device major from
	/proc/devices and creates entry in /dev.
	* thosts.example: fixed format string for target hosts.
	* install.sh: merged install.patch by marsus
	
2005-05-03  Paolo Ardoino  <ardoino.gnu@disi.unige.it>
	* ksb26-0.0.1 for Linux v2.6.x
	* ksb26.8.gz: added ksb26/ksb26manager man page (man 8 ksb26).
	* README: updated.
	* install.sh: updated [Creates also /etc/ksb26 directory which contains
		target hosts file].
	* kernel/*: removed kernel patch! ksb26 is a loadable kernel module.
	* user/ksb26manager.c: new userspace tool.

2005-04-17  Paolo Ardoino  <ardoino.gnu@disi.unige.it>
	* ksb26-2.6.11-2 for Linux v2.6.11
	* README: updated.
	* kernel/lkm/ksb26lkm.c: bug fix to character device handler (thanks
		Michel).
	* kernel/kpatch/ksb26.h: small bug fixes.
	* kernel/kpatch/ksb26.h: fixed handle of SOCKS's working status.
	* kernel/kpatch/ksb26.h, kernel/lkm/ksb26*: 
		ADDED SUPPORT FOR SOCKS v4. EXPERIMENTAL!!![read README]

2005-03-29  Paolo Ardoino  <ardoino.gnu@disi.unige.it>
	* kernel/kpatch/ksb26.c: code cleanup.
	* kernel/lkm/ksb26*: code cleanup.
	* kernel/lkm/ksb26*: new format for strings passed to ksb26 device.
	* README: updated.
	* install.sh: updated.

2005-03-21  Paolo Ardoino  <ardoino.gnu@disi.unige.it>
	* ksb26-2.6.11-1 for Linux v2.6.11
	* kernel/lkm/ksb26*: splitted kernel modules in different
		files.
	* kernel/lkm/ksb26*: code cleanup, fixed device handler.
	* kernel/lkm/ksb26_misc.h: fixed ksb26_isip(), ksb26_getline.
	* kernel/kpatch/ksb26.c: code cleanup.

2005-02-21  Paolo Ardoino  <ardoino.gnu@disi.unige.it>
	* ksb26-2.6.11 for Linux v2.6.11

2005-02-21  Paolo Ardoino  <ardoino.gnu@disi.unige.it>
	* ksb26-2.6.10.4 for Linux v2.6.10
	* kernel/kpatch/ksb26.c: support for chain of SOCKS5.
	* README: updated.

2005-02-21  Paolo Ardoino  <ardoino.gnu@disi.unige.it>
	* ksb26-2.6.10.3 for Linux v2.6.10
	* kernel/kpatch/ksb26.c, kernel/include/ksb26.h,
	  kernel/ksb26lkm.c: fixed semaphores, handle of ds/fs,
		circular double linked lists.
	* user/ksb26manager: now it clears SOCKS list without
		removing ksb26lkm.

2005-01-25  Paolo Ardoino  <ardoino.gnu@disi.unige.it>
	* ksb26-2.6.10.2 for Linux v2.6.10
	* kernel/kpatch/ksb26.c sys_connect_ksb26(), 
	  kernel/ksb26lkm.c modinit(),
	  kernel/kpatch/ksb26.h :
	  	circular double linked list for SOCKS list.

2005-01-24  Paolo Ardoino  <ardoino.gnu@disi.unige.it>
	* ksb26-2.6.10.1 for Linux v2.6.10
	* kernel/kpatch/ksb26.c sys_connect_ksb26(), 
	  kernel/ksb26lkm.c modinit() :
		fixed handle of semaphore ( ksb26_sem ).

2005-01-21  Paolo Ardoino  <ardoino.gnu@disi.unige.it>
	* ksb26-2.6.10 for Linux v2.6.10
	* kernel/kpatch/ksb26.c sys_connect_ksb26(): 
		fixed browser problem, handling sockets' flags.

2004-10-24  Paolo Ardoino  <ardoino.gnu@disi.unige.it>
	* README fixed

2004-10-20  Paolo Ardoino  <ardoino.gnu@disi.unige.it>
	* First public release of ksb26 ( for Linux v2.6.9 )



install.sh
#***************************************************************************#
# Kernel SOCKS Bouncer (install.sh) for 2.6.x kernels [ksb26]               #
# (c) 2004-2005 Paolo Ardoino <ardoino.gnu@disi.unige.it>                   #
#***************************************************************************#
#									    #
# This program is free software; you can redistribute it and/or modify	    #
# it under the terms of the GNU General Public License as published by	    # 
# the Free Software Foundation; either version 2 of the License, or	    # 
# (at your option) any later version.					    #
# This program is distributed in the hope that it will be useful,	    #
# but WITHOUT ANY WARRANTY; without even the implied warranty of	    #
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the             #
# GNU General Public License for more details.				    #
#									    #
# You should have received a copy of the GNU General Public License	    #
# along with this program; if not, write to the Free Software		    #
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA #
#***************************************************************************#


installer_version=0.0.2
kernel_dir=/usr/src/linux
bin_dir=/usr/bin

echo "Kernel Socks Bouncer for 2.6.x kernels Installer v$installer_version"
echo ""

# Checking all components of ksb26
echo "Checking components of ksb26:"
if [ ! -d kernel ] || [ ! user ] ; then
	echo "Cannot find subdirectoryes containing components of ksb26"
	exit -1
fi
echo -n "kernel/ksb26.c -> "
if [ ! -e kernel/ksb26.c ] ; then
	echo "not found!"
	exit -1
else
	echo "found!"
fi
echo -n "user/ksb26manager.c -> "
if [ ! -e user/ksb26manager.c ] ; then
	echo "not found!"
	exit -1
else
	echo "found!"
fi

# Checking Linux kernel source directory
if [ ! -d $kernel_dir ] || [ ! $kernel_dir/net ] || [ ! $kernel_dir/include ]  || [ ! $kernel_dir/kernel ] ; then
	echo "Cannot find Linux Kernel source in $kernel_dir"
	echo "Please edit this script and set 'kernel_dir' to the correct value'"
	exit -1
fi

# Starting installation
echo "Building kernel/ksb26.c"
make -C $kernel_dir/ SUBDIRS=$PWD/kernel V=1 modules
if [ ! -e kernel/ksb26.ko ] ; then
	exit -1
fi
echo "Building user/ksb26manager.c"
make -C user/
if [ ! -e user/ksb26manager ] ; then
	exit -1
fi
echo "Installing kernel/ksb26.ko"
make -C $kernel_dir/ SUBDIRS=$PWD/kernel V=1 modules_install
depmod
echo "Installing user/ksb26manager"
cp user/ksb26manager $bin_dir
if [ ! -e $bin_dir/ksb26manager ] ; then
	exit -1
fi
echo "Creating /etc/ksb26 directory"
mkdir /etc/ksb26
if [ ! -d /etc/ksb26 ] ; then
	exit -1
fi
echo "Installing configuration files"
cp thosts.example /etc/ksb26/
cp socks.example /etc/ksb26/
if [ ! -e /etc/ksb26/thosts.example ] ; then
	exit -1
fi
if [ ! -e /etc/ksb26/socks.example ] ; then
	exit -1
fi
echo "Installing man page (man 1 ksb26)"
cp ksb26.1.gz /usr/share/man/man1/
if [ ! -e /usr/share/man/man1/ksb26.1.gz ] ; then
	exit -1
fi
echo "Done"
echo ""
echo "Now you have to :"
echo "* read man page (man ksb26) for informations about the use of ksb26" 
echo "* set target hosts in /etc/ksb26/thosts"


kernel/ksb26.c
/*****************************************************************************/
/* Kernel SOCKS Bouncer (Loadable Kernel Module) for 2.6.x kernels [KSB26]   */
/* (c) 2004-2008 Paolo Ardoino <paolo.ardoino@gmail.com>                     */
/*****************************************************************************/
/*									     */
/* This program is free software; you can redistribute it and/or modify	     */
/* it under the terms of the GNU General Public License as published by	     */
/* the Free Software Foundation; either version 2 of the License, or	     */
/* (at your option) any later version.					     */
/* This program is distributed in the hope that it will be useful,	     */
/* but WITHOUT ANY WARRANTY; without even the implied warranty of	     */
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the	     */
/* GNU General Public License for more details.				     */
/*									     */
/* You should have received a copy of the GNU General Public License	     */
/* along with this program; if not, write to the Free Software		     */
/* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA */
/*****************************************************************************/

#include "ksb26.h"

static int unpatch_unix_stream_connect(void)
{
	if (unix_stream_ops && orig_unix_stream_connect) {
		unix_stream_ops->connect = orig_unix_stream_connect;
		return 0;
	}
	return -1;
}

static int patch_unix_stream_connect(void)
{
	struct socket *sock_stream = NULL;

	if (sock_create(2, 1, 0, &sock_stream) < 0)
		return -1;
        if (sock_stream && (unix_stream_ops = sock_stream->ops)) {
		orig_unix_stream_connect = unix_stream_ops->connect;
		unix_stream_ops->connect = ksb26_unix_stream_connect;
		sock_release(sock_stream);
	}
	return 0;
}

static int device_open(struct inode *inode, struct file *file)
{
	if(ksb26_dev_bool == 1) 
		return -EBUSY;	
	ksb26_dev_bool++;
	try_module_get(THIS_MODULE);
	return 0;
}

static int device_release(struct inode *inode, struct file *file)
{
	ksb26_dev_bool--;
	module_put(THIS_MODULE);
	return 0;
}

static ssize_t device_write(struct file *filp, const char *buffer, size_t lenght, loff_t *offset)
{
	int bytes_write = 0, linelen = 0;
	char *aux, *line;
	
	for(bytes_write = 0; bytes_write < lenght && bytes_write < KSB26_IOCTL_BUFLEN; bytes_write++)
		get_user(ksb26_ioctl_msg[bytes_write], buffer + bytes_write);
	ksb26_ioctl_msg[lenght - 1] = '\0';
	aux = ksb26_ioctl_msg;
	down_interruptible(&ksb26_listsem);
	while(aux < ksb26_ioctl_msg + lenght) {
		if((linelen = ksb26_getline(aux, -1, '\n', &line)) == 0) {
			up(&ksb26_listsem);
			return bytes_write;
		}
                printk("%s\n", line);
		if(line[0] == 'S')
			if(ksb26_add_socks(line) == 0)
				printk("[%s] Warning: failed to add SOCKS in '%s' line.\n", MODNAME, line);
		if(line[0] == 'H')
			if(ksb26_add_thost(line) == 0)
				printk("[%s] Warning: failed to add host in '%s' line.\n", MODNAME, line);
		if(line[0] == 'C') {
			ksb26_clear();
			printk("[%s] SOCKS list cleaned.\n", MODNAME);
		}
		aux = aux + linelen + 1;
		kfree(line);
	}
	up(&ksb26_listsem);
	return bytes_write;
}

static int __init modinit(void)
{
	printk("[%s v%s] MODULE LOADED\n", MODNAME, KSB26_VERSION);
	if((ksb26_major = register_chrdev(0, KSB26_DEV_NAME, &fops)) < 0) {
		printk("[%s] Cannot open device %s with %d major.\n", MODNAME, KSB26_DEV_NAME, ksb26_major);
		return -ENODEV;
	}
	printk("[%s] Device %s opened.\n", MODNAME, KSB26_DEV_NAME);
	printk("[%s] Major device number = %d.\n", MODNAME, ksb26_major);
	printk("[%s] mknod %s c %d 0\n", MODNAME, KSB26_DEV_NAME, ksb26_major);
	KSB26_INIT_LIST_HEAD(&ksb26_slh);
	KSB26_CLEAR_HSTRUCT(ksb26_slh);
	KSB26_INIT_LIST_HEAD(&ksb26_bhlh);
	KSB26_CLEAR_HSTRUCT(ksb26_bhlh);
	ksb26_connect_cnt = 0;
	ksb26_nsocks = 0;
	sema_init(&ksb26_listsem, 1);
	sema_init(&ksb26_cntsem, 1);
	printk("[%s] Hijacking SYS_CONNECT syscall.\n", MODNAME);
	ksb26_maxsocks = nsocks;
	printk("[%s] ksb26_maxsocks = %d\n", MODNAME, ksb26_maxsocks);
	if(patch_unix_stream_connect() == -1) {
		printk("[%s] Warning: failed to patch unix_stream_connect.\n", MODNAME);
		return -1;
	}
	printk("[%s] unix_stream_connect patched successfully.\n", MODNAME);
	return 0;
}

static void __exit modcleanup(void)
{
	unregister_chrdev(ksb26_major, KSB26_DEV_NAME);
        printk("[%s] Cannot unregister device %s.\n", MODNAME, KSB26_DEV_NAME);
	
        if(unpatch_unix_stream_connect() == -1) {
		printk("[%s] Warning: failed to unpatch unix_stream_connect.\n", MODNAME);
		return;
	}
	printk("[%s] unix_stream_connect unpatched successfully.\n", MODNAME);
	printk("[%s] MODULE CLEANED\n", MODNAME);
	return;
}

module_init(modinit);
module_exit(modcleanup);

module_param(nsocks, int, 0);

MODULE_AUTHOR("Paolo Ardoino");
MODULE_DESCRIPTION("Kernel Socks Bouncer for 2.6.x kernels\n");
MODULE_LICENSE("GPL");


kernel/ksb26.h
/*****************************************************************************/
/* Kernel SOCKS Bouncer (Loadable Kernel Module) for 2.6.x kernels [KSB26]   */
/* (c) 2004-2008 Paolo Ardoino <paolo.ardoino@gmail.com>                     */
/*****************************************************************************/
/*									     */
/* This program is free software; you can redistribute it and/or modify	     */
/* it under the terms of the GNU General Public License as published by	     */
/* the Free Software Foundation; either version 2 of the License, or	     */
/* (at your option) any later version.					     */
/* This program is distributed in the hope that it will be useful,	     */
/* but WITHOUT ANY WARRANTY; without even the implied warranty of	     */
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the	     */
/* GNU General Public License for more details.				     */
/*									     */
/* You should have received a copy of the GNU General Public License	     */
/* along with this program; if not, write to the Free Software		     */
/* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA */
/*****************************************************************************/

#include <linux/syscalls.h>
#include <linux/init.h>
#include <linux/module.h>
#include <linux/kernel.h>
#include <linux/moduleparam.h>
#include <asm/unistd.h>
#include <asm/uaccess.h>
#include <linux/string.h>
#include <linux/in.h>
#include <asm/semaphore.h>
#include <linux/byteorder/generic.h>
#include <net/sock.h>
#include <linux/aio.h>
#include <linux/inet.h>
#include <linux/slab.h>
#include <linux/preempt.h>
#include <linux/un.h>
#include <linux/types.h>
#include <linux/fs.h>
#include <linux/file.h>
#include <net/af_unix.h>
#include <linux/net.h>

#define MODNAME "ksb26"
#define KSB26_VERSION "0.0.1"

#define KSB26_DEV_NAME "ksb26"
#define KSB26_IOCTL_BUFLEN 4096
#define KSB26_MAXSOCKS 1

#define KSB26_INIT_LIST_HEAD(ptr) do { \
	(ptr)->next = (ptr); (ptr)->prev = (ptr); \
} while(0)
#define KSB26_CLEAR_HSTRUCT(ksb26_hs) do { \
	memset(ksb26_hs.ip, '\0', 18); \
	memset(ksb26_hs.ipc, '\0', 4); \
	ksb26_hs.port = 0; \
	ksb26_hs.naddr = 0; \
	ksb26_hs.nport = 0; \
	ksb26_hs.wrk = 0; } while(0)

#define DEBUG 1

unsigned int ksb26_nsocks;
unsigned int ksb26_nbhddrs;
struct ksb26_host ksb26_slh;
struct ksb26_host ksb26_bhlh;
struct semaphore ksb26_listsem;
struct semaphore ksb26_cntsem;
int ksb26_connect_cnt; 
int ksb26_maxsocks;

static int nsocks = KSB26_MAXSOCKS;
static int ksb26_major;
static char ksb26_ioctl_msg[KSB26_IOCTL_BUFLEN];
static int ksb26_dev_bool = 0;

static int device_open(struct inode *, struct file *);
static int device_release(struct inode *, struct file *);
static ssize_t device_write(struct file *, const char *, size_t, loff_t *);


static struct file_operations fops = {
	.write = device_write,
	.open = device_open,
	.release = device_release
};

struct ksb26_host {
	unsigned long naddr;
	unsigned short nport;
	char ip[18];
	char ipc[5];
	int port;
	int wrk;
	struct ksb26_host *next;
	struct ksb26_host *prev;
};

static struct proto_ops *unix_stream_ops = NULL;
static int (*orig_unix_stream_connect)(struct socket *sock, struct sockaddr *uaddr, int addr_len, int flags) = NULL;
static int ksb26_unix_stream_connect(struct socket *sock, struct sockaddr *uaddr, int addr_len, int flags);
asmlinkage long sys_send(int fd, void __user * buff, size_t len, unsigned flags);

#include "ksb26_misc.h"
#include "ksb26_list.h"
#include "ksb26_core.h"



user/ksb26manager.c
/*****************************************************************************/
/* Kernel SOCKS Bouncer Manager for 2.6.x kernels [KSB26]                    */
/* (c) 2004-2008 Paolo Ardoino <paolo.ardoino@gmail.com>                     */
/*****************************************************************************/
/*									     */
/* This program is free software; you can redistribute it and/or modify	     */
/* it under the terms of the GNU General Public License as published by	     */
/* the Free Software Foundation; either version 2 of the License, or	     */
/* (at your option) any later version.					     */
/* This program is distributed in the hope that it will be useful,	     */
/* but WITHOUT ANY WARRANTY; without even the implied warranty of	     */
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the	     */
/* GNU General Public License for more details.				     */
/*									     */
/* You should have received a copy of the GNU General Public License	     */
/* along with this program; if not, write to the Free Software		     */
/* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA */
/*****************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define PNAME "ksb26manager"
#define VERSION "0.0.1"


#define PROC_DEV "/proc/devices"
#define DEFAULT_THOSTS "/etc/ksb26/thosts"
#define DEFAULT_SOCKS "/etc/ksb26/socks"
#define DEFAULT_LKM "ksb26"
#define DEFAULT_DEV "/dev/ksb26"
#define DEFAULT_WTIME 600
#define DEFAULT_NSOCKS 2
#define MAX_SOCKS 10

#define BUFFER_SIZE 1024
#define VARSIZE 128

int nsocks, wtime;
char ksb26_lkm[VARSIZE], ksb26_dev[VARSIZE];
char thosts[VARSIZE], socks[VARSIZE];

void help(void)
{
    printf("[%s] Usage: %s [-h] [-d ksb26_device] [-m ksb26_lkm_name]\n", PNAME, PNAME);
    printf("\t[-n nsocks] [-t thosts_file] [-w wait_time]\n\n");
    printf("\t[-d ksb26_device]: character device used by ksb26 [default:%s]\n", DEFAULT_DEV);
    printf("\t[-m ksb26_lkm_name]: lkm's name to use in modprobe command [default:%s]\n", DEFAULT_LKM);
    printf("\t[-n nsocks]: number of SOCKS used in chains [default:%d]\n", DEFAULT_NSOCKS);
    printf("\t[-t thosts_file]: file containing target hosts [default:%s]\n", DEFAULT_THOSTS);
    printf("\t[-s socks_file]: file containing default socks list [default:%s]\n", DEFAULT_SOCKS);
    printf("\t[-w wait_time]: number of seconds to wait before updating the SOCKS list [default:%d]\n", DEFAULT_WTIME);
    printf("\t[-h]: help\n");
    printf("Please read man page (man ksb26) or README to learn format of target hosts and SOCKS\n");
}

int get_major(char *name)
{
    FILE *fp;
    char tbuf[BUFFER_SIZE];
    int major = -1;

    if(!(fp = fopen(PROC_DEV, "r"))) {
        fprintf(stderr, "[%s] error: fopen: cannot find '%s' file\n", PNAME, PROC_DEV);
        return -1;
    }
    while(fgets(tbuf, BUFFER_SIZE, fp)) {
        if(strstr(tbuf, name))
            major = atoi(tbuf);
    }
    return major;
}

int insmod_ksb26(void)
{
    char cmd[BUFFER_SIZE];
    char buff[BUFFER_SIZE];
    FILE *fp;
    int major;

    snprintf(cmd, BUFFER_SIZE, "rmmod %s 1&> /dev/null", ksb26_lkm);
    system(cmd);
    printf("%s\n", cmd);
    snprintf(cmd, BUFFER_SIZE, "rm %s", ksb26_dev);
    system(cmd);
    printf("%s\n", cmd);
    snprintf(cmd, BUFFER_SIZE, "modprobe %s nsocks=%d", ksb26_lkm, nsocks);
    if(!(fp = popen(cmd, "r"))) {
        fprintf(stderr, "[%s] error: popen: failed popen(\"%s\", \"r\")\n", PNAME, cmd);
        return -1;
    }
    while(fgets(buff, BUFFER_SIZE, fp)) {
        if(strstr(buff, "FATAL") || strstr(buff, "fatal") || strstr(buff, "Fatal")) {
            fprintf(stderr, "[%s] error: modprobe: %s", PNAME, buff);
            pclose(fp);
            return -1;
        }
    }
    pclose(fp);
    if((major = get_major(ksb26_lkm)) < 0) {
        printf("[%s] error: cannot find '%s' major\n", PNAME, ksb26_lkm);
        return -1;
    }
    printf("major = %d\n", major);
    snprintf(cmd, BUFFER_SIZE, "mknod %s c %d 0", ksb26_dev, major);
    system(cmd);
    return 0;
}

int load_thosts_list(void)
{
    char cmd[BUFFER_SIZE];
    FILE *fp_dev, *fp_thosts;
    char buff[BUFFER_SIZE];

    if(!(fp_thosts = fopen(thosts, "r"))) {
        fprintf(stderr, "[%s] error: fopen: cannot find '%s' file\n", PNAME, thosts);
        return -1;
    }
    if(!(fp_dev = fopen(ksb26_dev, "w"))) {
        fprintf(stderr, "[%s] error: fopen: cannot open '%s' device\n", PNAME, ksb26_dev);
        fclose(fp_thosts);
        return -1;
    }
    while(fgets(buff, BUFFER_SIZE, fp_thosts)) {
        if(buff[0] != '#') {
            snprintf(cmd, BUFFER_SIZE, "%s\n", buff);
            fputs(cmd, fp_dev);
        }
    }
    fclose(fp_thosts);
    fclose(fp_dev);
    return 0;
}

int load_socks_list(void)
{
    char cmd[BUFFER_SIZE];
    FILE *fp_dev, *fp_socks;
    char buff[BUFFER_SIZE];

    if(!(fp_socks = fopen(socks, "r"))) {
        fprintf(stderr, "[%s] error: fopen: cannot find '%s' file\n", PNAME, socks);
        return -1;
    }
    if(!(fp_dev = fopen(ksb26_dev, "w"))) {
        fprintf(stderr, "[%s] error: fopen: cannot open '%s' device\n", PNAME, ksb26_dev);
        fclose(fp_socks);
        return -1;
    }
    while(fgets(buff, BUFFER_SIZE, fp_socks)) {
        if(buff[0] != '#') {
            snprintf(cmd, BUFFER_SIZE, "%s\n", buff);
            fputs(cmd, fp_dev);
        }
    }
    fclose(fp_socks);
    fclose(fp_dev);
    return 0;
}

int clear(void)
{
    char cmd[BUFFER_SIZE];
    FILE *fp_dev;
    char buff[BUFFER_SIZE];

    snprintf(cmd, BUFFER_SIZE, "C\n", ksb26_dev);
    if(!(fp_dev = fopen(ksb26_dev, "w"))) {
        fprintf(stderr, "[%s] error: fopen: cannot open '%s' device\n", PNAME, ksb26_dev);
        fclose(fp_dev);
        return -1;
    }
    fputs(cmd, fp_dev);
    fclose(fp_dev);
    return 0;
}

int main(int argc, char *argv[])
{
    int i;

    nsocks = DEFAULT_NSOCKS;
    wtime = DEFAULT_WTIME;
    strncpy(ksb26_lkm, DEFAULT_LKM, VARSIZE - 1);
    strncpy(ksb26_dev, DEFAULT_DEV, VARSIZE - 1);
    strncpy(thosts, DEFAULT_THOSTS, VARSIZE - 1);
    strncpy(socks, DEFAULT_SOCKS, VARSIZE - 1);
    printf("\n");
    printf("[%s] Kernel SOCKS Bouncer Manager v%s for ksb26\n", PNAME, VERSION);
    printf("[%s] by Paolo Ardoino <paolo.ardoino@gmail.com>\n\n", PNAME);
    for(i = 0; i < argc; i++) {
        if(strcmp(argv[i], "-h") == 0 || strcmp(argv[i], "--help") == 0) {
            help();
            exit(EXIT_SUCCESS);
        }
    }
    for(i = 0; i < argc; i++) {
        if(strcmp(argv[i], "-n") == 0 && argv[i + 1]) {
            nsocks = atoi(argv[i + 1]);
            if(nsocks <= 0 || nsocks > 10) {
                fprintf(stderr, "[%s] error: nsocks: invalid value\n", PNAME);
                exit(EXIT_FAILURE);
            }
        }
        if((strcmp(argv[i], "-m") == 0) && argv[i + 1]) {
            memset(ksb26_lkm, '\0', VARSIZE);
            strncpy(ksb26_lkm, argv[i + 1], VARSIZE - 1);
        }
        if(strcmp(argv[i], "-t") == 0 && argv[i + 1]) {
            memset(thosts, '\0', VARSIZE);
            strncpy(thosts, argv[i + 1], VARSIZE - 1);
        }
        if(strcmp(argv[i], "-s") == 0 && argv[i + 1]) {
            memset(socks, '\0', VARSIZE);
            strncpy(socks, argv[i + 1], VARSIZE - 1);
        }
        if(strcmp(argv[i], "-d") == 0 && argv[i + 1]) {
            memset(ksb26_dev, '\0', VARSIZE);
            strncpy(ksb26_dev, argv[i + 1], VARSIZE - 1);
        }
        if(strcmp(argv[i], "-w") == 0 && argv[i + 1]) {
            wtime = atoi(argv[i + 1]);
            if(wtime <= 0) {
                fprintf(stderr, "[%s] error: wtime: invalid value\n", PNAME);
                exit(EXIT_FAILURE);
            }
        }
    }
    printf("[%s] modprobe %s nsocks=%d\n", PNAME, ksb26_lkm, nsocks);
    printf("[%s] ksb26 device: %s\n", PNAME, ksb26_dev);
    printf("[%s] thosts file: %s\n", PNAME, thosts);
    printf("[%s] socks file: %s\n", PNAME, socks);
    printf("[%s] wait time: %d\n", PNAME, wtime);

    if(insmod_ksb26() == -1) exit(EXIT_FAILURE);
    if(load_thosts_list() == -1) exit(EXIT_FAILURE);
    while(1) {
        if(clear() == -1) exit(EXIT_FAILURE);
        if(load_socks_list() == -1) exit(EXIT_FAILURE);
        sleep(wtime);
    }
    return 0;
}

